### DISCUSSION TITLE
Immutable public properties (read-only/constants after Livewire initial mounted / dehydrated)

### DISCUSSION THREAD

**Discussion by camya** [livewire/livewire#4614] [3. Feature ideas]
---
2022-02-08 10:33 - Updated post (See original idea post below)
---

@lennardv2 posted a gist for the Trait idea (WithProtection / WithPropertyProtection / WithMethodProtection): 

> "This trait is used to protect public livewire properties from being changed by the user by default in frontend via Livewire JS calls. If you want to 'expose' methods and properties you must use the $callable and $mutable properties to do so. ... By default every property defined in $rules or rules() is considered mutable. This trait will throw an exception if mutable and callable properties are non-existant on your component; ..." 

**"WithProtection" Trait Gist / WIP:** https://gist.github.com/lennardv2/9feec1028289cdb3c7cb1033e87442f0

---

Obsolete  2022-02-07 13:39 - Updated post (See original idea post below)

---

This feature allows defining a list of public properties of a Livewire component, which should be **immutable (read-only)**. 
A modification of these properties **leads to a Livewire Exception**.

```
use WithImmutableProps;

public $supportedComponentLocations = ['content', 'sidebar'];

protected array $immutableProps = [
    'supportedComponentLocations',
];
```

**Why is this needed?**  

In a Livewire component, there are often "public properties", which nevertheless shouldn't be changed under any circumstances. 

Examples are $componentActsAs = 'frontend' or $supportedComponentLocations = ['content', 'sidebar']. 

Unfortunately, every public property in Livewire could be changed from the JavaScript console without any problem.

Example: `Livewire.all()[0].supportedComponentLocations = ['another-location']`

**Possible solution:** @lennardv2 created a **WithImmutableProps** Trait which works great. I updated a bit in order to use it also which class inheritance. (for Base Livewire Components for example) 

**I hope a feature / trait like this will be added to the Livewire core**. Maybe there are other ideas how to implement this feature. Any idea is welcome. ;)

**WithImmutableProps trait**

```
<?php

/**
 * Livewire Trait "with immutable properties"
 *
 * Define which of the Livewire components public properties should be handled as immutable.
 *
 * Any modification of these immutable properties will lead to an exception.
 *
 * Example in a Livewire component:
 *
 * public $supportedComponentLocations = ['content', 'sidebar'];
 *
 * // Set the list using a property:
 * protected array $immutableProps = [
 *     'supportedComponentLocations',
 * ];
 *
 * // Alternative: Set the list using a method:
 * // This allows to merge immutable props, e.g. in case of class inheritance.
 * protected function immutableProps(): array
 * {
 *     return [
 *         'supportedComponentLocations',
 *     ];
 * }
 */
trait WithImmutableProps
{
    /**
     * Throw an exception if and immutable property is updating. (Livewire lifecycle hook "updating")
     *
     * @throws \Exception
     */
    public function updatingWithImmutableProps(string $propertyName): void
    {
        if (method_exists($this, 'immutableProps')) {

            $props = $this->immutableProps();

        } elseif (isset($this->immutableProps)) {

            $props = $this->immutableProps;

        }

        if (empty($props)) {
            return;
        }

        if (in_array($propertyName, $props)) {
            throw new \Exception("You cannot change the value of the immutable property '{$propertyName}'");
        }
    }
}
```

**Example Livewire Component A: Use protected array $immutableProps = [];** 

```
use WithImmutableProps;

public $supportedComponentLocations = ['content', 'sidebar'];

protected array $immutableProps = [
    'supportedComponentLocations',
];
```

The follwing will throw an exception:

`Livewire.all()[0].supportedComponentLocations = ['another-location']`

_"You cannot change the value of the immutable property 'supportedComponentLocations'"_

**Example Livewire Component B: Use protected function immutableProps(): array {}** 

It's the same like $immutableProps, but it allows to merge immutable props in case of class inheritance.

```
use WithImmutableProps;

public $supportedComponentLocations = ['content', 'sidebar'];

protected function immutableProps(): array
{
    return [
        'supportedComponentLocations',
    ];
}

// or ...
protected function immutableProps(): array
{
    return array_merge(
        parent::immutableProps(),
        [
            'supportedComponentLocations',
        ]
    );
}
```

--- 

Obsolete: Original feature request

---

This feature allows defining a list of public properties of a Livewire component, which should be **immutable (read-only)** after the Livewire component was mounted / dehydrated initially. The properties/values are stored in "wire:initial-data" for subsequent requests. (see implementation idea below). 

A modification of these properties (on subsequent requests) **leads to a Livewire Exception**.

**Why I think this feature is needed, in my opinion?**

If I re-use a Livewire components in different locations of my application, I often have to pass unchangeable initial values (kind of constants) to the component in order to configure its behavior at the specific location. 

`<livewire:product :productId="$productId" componentActsAs="frontend" componentLocation="sidebar" />`

(**Examples**: Show component in Frontend/Backend, e.g **$componentActsAs**="frontend", Insert component in Sidebar-Slot/Content-Slot e.g. **$componentLocation**="sidebar") 

More use cases are public variables like $withPaidContent or $productId which are initialized on mount(), but which should never be changable under any circumstances.

**These values should be read-only** in these use cases. If a user tampers with this data after initialization (from code or from the JavaScript Console), Livewire should throw an Exception. (Check via checksum, see implementation idea below)

**Current problem. Tampering with public properties is easy.**

At the moment, any user with a bit of Livewire & JS knowledge can tamper with the variables mentioned above easily via Chrome JavaScript Console. (E.g.  Livewire.all()[0].withPaidContent = true, Livewire.all()[0].componentActsAs = 'backend', Livewire.all()[0].productId = 321, Livewire.all()[0].componentLocation ='sidebar') 

The immutableMountProperties feature will prevent this.

**Idea:**

It should be possible to set a list of "immutable properties" in the Livewire component via 

`$immutableMountProperties = []`

The example below sets the folowing immutable values:

```
productId = 111
componentActsAs = 'frontend'
componentLocation = 'content'
supportedComponentLocations = ['content', 'sidebar']
withPaidContent = false 
```

```
# In Blade template

<livewire:product productId="111" componentActsAs="frontend" />

# In Livewire component

class Product extends Component
{
    // These properties are marked as immutable in $immutableMountProperties
    public $productId;
    public $componentActsAs;
    public $componentLocation;
    public $supportedComponentLocations = ['content', 'sidebar'];
    public $withPaidContent;

    // Just a regular public property
    public $searchQuery = '';

    // List public properties which should be immutable after mount. (Read-only)
    protected $immutableMountProperties = [
        'productId',
        'componentActsAs', 
        'componentLocation',
        'supportedComponentLocations',
	'withPaidContent',
    ];

    public function mount($productId, $componentActsAs = 'frontend', $componentLocation = 'content')
    {
        $this->productId = $productId;
        $this->componentActsAs = $componentActsAs;
        $this->componentLocation = $componentLocation;

        // In this example, activeSubscription() returns false
        $this->withPaidContent = User::query()->activeSubscription();
    }
}
```

**Loose implementation idea:**

1) **Initial Setup in Livewire Component / PHP**: 
    - Define a list of public properties which should be immutable. (see above, $immutableMountProperties)
    - On initial dehydration of the component it generates the **wire:initial-data** JSON output in Blade template with immutable fields, values, and special immutable **checksum**.

2) **Lifecycle / Validation (Livewire Ajax request + Livewire Component / PHP)**: 
    - "**wire:initial-data**" contains the list of immutable marked public properties and their initial values + a checksum for these fields/values.
    - On subsequent Livewire requests, check if the fields and checksum match the fields & checksum stored in the wire:initial-data.
    - Livewire **throws an exception** if immutable properties are modified. 

**Example of the "wire:initial-data" JSON with "dataImmutable" & "checksumImmutable"**

`<div wire:id="KEX81SSK3fS1df5PdazX" wire:initial-data="{&quot;fingerprint&quot;:{ ... }}">`

In the example below, the searchQuery is the only public property not marked as immutable.

**Livewire should throw an Exception** if the data in "dataImmutable" was modiefied on a subsequent Ajax requests.

```
{
    "fingerprint": {
        "id": "KEX81SSK3fS1df5PdazX",
    },
    "serverMemo": {

        "dataImmutable": {
            "productId": 111,
            "componentActsAs": "frontend", 
            "componentLocation": "content",  
            "supportedComponentLocations": ["content", "sidebar"], 
            "withPaidContent": false
        },
        "checksumImmutable": "abcd1234abcd1234abcd1234abcd1234abcd1234abcd1234abcd1234",

        "data": {
            "searchQuery": "",
        },
        "checksum": "f8228c638794ddf0431905bea6775c5d7d111b549f81d20bf05e3257b179933b"

    }
}
```

**Better / other ideas to prevent modifications of public properties?**

This is **only a brain dump Idea**, and I'm not really sure, if it's really possible to implement it that way and I missed something essential.  

Maybe there are **other ideas to prevent users from tampering with these essential initial public property (kind of constant) values** of a Livewire component, which makes this idea obsolete. If so, I'm really interested in hearing about these solutions too.

**Comment by lennardv2** [comment]
Yes, i agree. Often you just want some configuration of a livewire component, without it being subject to changes from the user. For example: an upload component with configurable file types to be uploaded.

At first i was not aware of this. The docs do say that public props are always visible for a user (see: https://laravel-livewire.com/docs/2.x/properties#important-notes), but not that they can be mutated too. I thought the checksum / rules() thing would protect this from happening.

**Comment by lennardv2** [comment]
This will probably work, what do you think @camya 

```php

trait WithImmutableProps
{
    public function updatingWithImmutableProps($key, $value)
    {
        if (!isset($this->immutableProps)) return;
        if (!is_array($this->immutableProps)) throw new Exception("immutableProps must be an array");

        if (in_array($key, $this->immutableProps)) {
            throw new Exception("You cannot change the value of the immutable property '{$key}'");
        }
    }
}
```

**Reply by camya** [reply]
Hi @lennardv2 

I guess, the whole immutable logic with checksum & tampering checks must be implemented inside Livewire's core.

It's required that the components "wire:initial-data" HTML attribute contains the array of "imutable" properties (from $immutableMountProperties) + the constant values from the initial dehydration of the component and a checksum. Any subsequent Ajax request validates this data against the properties set in PHP ($immutableMountProperties) and the checksum.

**Example of the "wire:initial-data" JSON with "dataImmutable" & "checksumImmutable"**

`<div wire:id="KEX81SSK3fS1df5PdazX" wire:initial-data="{&quot;fingerprint&quot;:{ ... }}">`

In the example below, the searchQuery is the only public property not marked as immutable.

**Livewire should throw an Exception** if the data in "dataImmutable" was modiefied on a subsequent Ajax requests.

```
{
    "fingerprint": {
        "id": "KEX81SSK3fS1df5PdazX",
    },
    "serverMemo": {

        "dataImmutable": {
            "productId": 111,
            "componentActsAs": "frontend", 
            "componentLocation": "content",  
            "supportedComponentLocations": ["content", "sidebar"], 
            "withPaidContent": false
        },
        "checksumImmutable": "abcd1234abcd1234abcd1234abcd1234abcd1234abcd1234abcd1234",

        "data": {
            "searchQuery": "",
        },
        "checksum": "f8228c638794ddf0431905bea6775c5d7d111b549f81d20bf05e3257b179933b"

    }
}
```

**Reply by lennardv2** [reply]
Hi @camya, why do you think it is needed to checksum the props? This trait just checks if someone wants to change the specified prop name. Seems to work on JS calls. For example ``Livewire.all()[0].withPaidContent = true`` will throw an exception as long as ``public/protected $immutableProps = ['otherProp', 'otherProp2'] `` isset on the component

**Reply by camya** [reply]
Hi @lennardv2. Nice. I tried your trait, and it indeed works. 

Totally missed that it's possible to use the [livecycle hooks](https://laravel-livewire.com/docs/2.x/traits) of Livewire inside a trait. (updatingWithImmutableProps).

Maybe my idea was just over-engineered :) and you trait is the solution for this problem. 

**I hope this feature / trait will be added to the Livewire core**, because I think, many people need it too.

Below you'll find an updated WithImmutableProps trait which now supports a immutableProps() method and a $immutableProps property.

**Update:** I moved the WithImmutableProps example code to the [original post](https://github.com/livewire/livewire/discussions/4614).

**Reply by lennardv2** [reply]
Great!

> **I hope this feature / trait will be added to the Livewire core**, because I think, many people need it too.

Yes, also I think I am not the only one who missed that public properties can be modified by the client. 

I myself made a trait that turns the logic around. Props that are specified in ``$mutableProps``  can be modified, others cant. It makes an exception for everything defined in rules(), because they must be mutable and have validation on them. Also I did the same thing for methods that accept arguments with a ``$callableMethods``.  

I made these properties public so they can be specified when using the component via @livewire.

And now it works for me like I thought it did in first place.

**Comment by lennardv2** [comment]
Here is a gist of the idea:

https://gist.github.com/lennardv2/9feec1028289cdb3c7cb1033e87442f0

**Reply by camya** [reply]
Nice. Just tried your implementation and it works like expected. 

Can you  additionally allow **methods, e.g. mutable() & callable()**, to set the mutable & callable array? 

This helps to merge them when using class inheritance. (For BaseModals, BaseComponents, ...)

```
// In BaseCompent

use WithProtection;

protected function mutable() {
    return [
        'baseComponentProperty'
    ];
}

protected function callable()  {
    return [
        'baseComponentMethod'
    ];
}
```

Than it's possible in the child class to merge them. (See below)

```
// In ChildCompent

// Now contains: baseComponentProperty, childComponentProperty
protected function mutable() 
{
    return array_merge(parent::mutable(), [
        'childComponentProperty'
    ]);
}
```

**Reply by lennardv2** [reply]
Ah yes, good point! I changed the trait a bit :) thanks

**Reply by camya** [reply]
@lennardv2 I updated the main post and added your gist on top.

**Comment by joshhanley** [comment]
@camya and @lennardv2 there was already a PR #3893 proposed for this.

We eventually decided that a `$readonly` property was best, but due to the breaking change, it's something that has to wait until V3 as Caleb didn't want to introduce another trait.

So your trait is a good solution for now. Can I suggest creating a new discussion just with the trait and putting it in the "Show and Tell" category so others can find it?

In regards to your comment @lennardv2 about thinking that the checksum protected properties from being modified. What the checksum does is prevents the state of the component from being modified outside of one of Livewire's valid methods.

By making sure the changes are controlled by Livewire, ensures that any changes run through the updating/ed hooks first, allowing you to run your own checks, validation, etc. to make sure the data the user has entered is correct (and they are authorised etc).

Hope this helps!

**Reply by lennardv2** [reply]
Hi @joshhanley,

Thanks for your reply. 

> In regards to your comment @lennardv2 about thinking that the checksum protected properties from being modified. What the checksum does is prevents the state of the component from being modified outside of one of Livewire's valid methods.

Yes that was an assumption on my part, but I also think this could be stated more clear in the docs. Here: https://laravel-livewire.com/docs/2.x/properties#important-notes

> Data stored in public properties is **made visible** to the front-end JavaScript. Therefore, you SHOULD NOT store sensitive data in them.

Should be something like: made visible **and is changeable**

And here: https://laravel-livewire.com/docs/2.x/security

@calebporzio I certainly would make the properties readonly by default in v3, even if it is a bit inconvenient. Because when you change it from readonly to writable it's a reminder that you need to take a look at the security side of things for this property.

**Reply by joshhanley** [reply]
@lennardv2 no worries! Yeah fair enough, PR's are welcome to the docs https://github.com/livewire/docs

Yeah that could be a good idea.

**Comment by camya** [comment]
> Can I suggest creating a new discussion just with the trait and putting it in the "Show and Tell" category so others can find it?

@lennardv2 Can you please add your code to a new post in the **"Show and Tell" category**? It's a great idea to show your solution to other people there. This post can be closed then.
